File: 0543-diameter-of-binary-tree/0543-diameter-of-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int dfs(TreeNode* root, int &answer) {
        if (!root) {
            return 0;
        }
        int leftMax = dfs(root->left, answer);
        int rightMax = dfs(root->right, answer);
        answer = max(answer, leftMax + rightMax);
        return max(leftMax, rightMax) + 1;
    }
    int diameterOfBinaryTree(TreeNode* root) {
        int answer = 0;
        dfs(root, answer);
        return answer;
    }
};

File: 0543-diameter-of-binary-tree/0543-diameter-of-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def __init__(self):
	    self.diameter = 0  # stores the maximum diameter calculated
        
    def depth(self, root):
        if root is None:
            return 0
        left, right = self.depth(root.left), self.depth(root.right)
        self.diameter = max(self.diameter, left + right) 
        return max(left, right) + 1
    
    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        self.depth(root)
        return self.diameter
        